staging/src/k8s.io/apimachinery/pkg/runtime/serializer/protobuf/protobuf.go

```diff
@@ -161,22 +162,35 @@ func (s *Serializer) Decode(originalData []byte, gvk *schema.GroupVersionKind, i
 	return unmarshalToObject(s.typer, s.creater, &actual, into, unk.Raw)
 }
 
+// EncodeWithAllocator writes an object to the provided writer.
+// In addition, it allows for providing a memory allocator for efficient memory usage during object serialization.
+func (s *Serializer) EncodeWithAllocator(obj runtime.Object, w io.Writer, memAlloc runtime.MemoryAllocator) error {
+	return s.encode(obj, w, memAlloc)
+}
+
 // Encode serializes the provided object to the given writer.
 func (s *Serializer) Encode(obj runtime.Object, w io.Writer) error {
+	return s.encode(obj, w, &runtime.SimpleAllocator{})
+}
+
+func (s *Serializer) encode(obj runtime.Object, w io.Writer, memAlloc runtime.MemoryAllocator) error {
 	if co, ok := obj.(runtime.CacheableObject); ok {
-		return co.CacheEncode(s.Identifier(), s.doEncode, w)
+		return co.CacheEncode(s.Identifier(), func(obj runtime.Object, w io.Writer) error { return s.doEncode(obj, w, memAlloc) }, w)
 	}
-	return s.doEncode(obj, w)
+	return s.doEncode(obj, w, memAlloc)
 }
 
-func (s *Serializer) doEncode(obj runtime.Object, w io.Writer) error {
+func (s *Serializer) doEncode(obj runtime.Object, w io.Writer, memAlloc runtime.MemoryAllocator) error {
+	if memAlloc == nil {
+		return fmt.Errorf("a memory allocator must be provided")
```

Comment:
I would prefer to logging an Error and falling back to SimpleAllocator in this case rather than failing the whole encoding.

Reply:
it cannot happen. I added it as a safety net, `doEncode` is an internal method.

Reply:
`memAlloc` is provided by `Encode` and `EncodeWithAllocator` methods.

Reply:
Yes - an one can call EncodeWithAllocator and pass nil memoryallocator

Reply:
In that case, a caller should use the `Encode` method or provide a noop allocator.
I would prefer to make it explicit as it is less error-prone, i.e. some middle layer doesn't pass the provided allocated down.

Reply:
They should use `Encode` but they made a bug. I think it's still better to mitigate the problem and not fail the request on the cost of worse performance.